### PR TITLE
Add support for listing tests (without running them).

### PR DESCRIPTION
**Pull request by stefanseefeld**
This patch adds support for listing tests without running them.
I occasionally run into test failures leading to the entire test run to fail, without a clear indication which test it was that caused the failure. Having the ability to list an (ordered) list of tests allows me to see the "current" test (i.e. the one following the last completed one) without hacking the code.

This isn't quite working yet, as the numba 'runtests.py' script invokes multiple instances of test suites, not all of them using a TestRunner. So this needs more work to be complete.
(I'm wondering whether this use-case is interesting enough in general for it to be submitted upstream.)

### CODE REVIEW

**Comment by stefanseefeld** [comment]
The last commit modifies the cudapy tests to not require jit-compilation at global scope, which would fail in the presence of cudadrv tests which reset the cuda context, thereby invalidating the pre-compiled functions.
(An alternative would be to insulate the cudadrv tests so they don't interfer with the (global) cuda context.)

File: numba/tests/__init__.py [hunk 1]
```diff
@@ -82,22 +92,59 @@ def _getParentArgParser(self):
...
+                              not self.test.countTestCases())):
+            self._do_discovery([])
+            
         if self.verbosity <= 0:
             # We aren't interested in informational messages / warnings when
             # running with '-q'.
             self.buffer = True
-
+            
+    def _do_discovery(self, argv, Loader=None):
```
Comment (pitrou) [inline numba/tests/__init__.py]: argv and Loader aren't used anywhere, right?

File: numba/testing.py [hunk 2]
```diff
@@ -49,49 +47,15 @@ def run_tests(suite, xmloutput=None, verbosity=1, nomultiproc=False):
...
-    - descriptions
-    - verbosity
-    - buffer
-    - failfast
-    - xmloutput [str]
-        Path of XML output directory
-    """
-    from numba import cuda
-
-    suite = discover_tests("numba.tests")
```
Comment (pitrou) [inline numba/testing.py]: It seems discover_tests is now unused, or is it still used elsewhere?

File: numba/tests/__init__.py [hunk 3]
```diff
@@ -82,22 +92,59 @@ def _getParentArgParser(self):
...
+        """The discovery process is complicated by the fact that:
+
+        * different test suites live under different directories
+        * some test suites may not be available (CUDA)
+        * some tests may have to be run serially, even in the presence of the '-m' flag."""
+
+        from numba import cuda
+        loader = unittest.TestLoader()
+        topdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
+        base_tests = loader.discover('%s/numba/tests'%topdir, 'test*.py', topdir)
```
Comment (pitrou) [inline numba/tests/__init__.py]: `os.path.join` should be preferred to %-formatting when joining paths.

File: numba/testing.py (hunk 2 above)
Comment (stefanseefeld) [inline numba/testing.py]: There are still various 'runtests.py' scripts, notably in the cuda test suites. I'm not sure how relevant these scripts are at this point, i.e. whether anyone would run them explicitly.

This pull request should in fact render them obsolete, so we may remove them as part of a 'cleanup' follow-up patch.

File: numba/tests/__init__.py (hunk 3 above)
Comment (stefanseefeld) [inline numba/tests/__init__.py]: Fair enough. How portable is the "numba/tests" notation these days, anyhow ? I thought all OSes understood that by now ? Or should I also use 'join' for those ?

File: numba/tests/__init__.py (hunk 1 above)
Comment (stefanseefeld) [inline numba/tests/__init__.py]: They aren't, but in fact, they should be (for completeness). 
With Python 2.x there was an explicit 'discover' argument, which for Python 3.x seems implicit.
With my current patch the `python -m numba.testing discover a b c` would ignore 'a', 'b' and 'c', but would work fine if spelled `python -m numba.testing a b c`.

The unittest code (and logic) is quite messy, with many different ways to solve the same problem. I'm still reverse-engineering bits and pieces to make sure I'm not altering or removing functionality in my patches.

File: numba/tests/__init__.py (hunk 3 above)
Comment (stefanseefeld) [inline numba/tests/__init__.py]: (I have substituted the '%s/...' formatting by 'join', as requested, but left the '.../...' paths intact.)

Comment (pitrou) [inline numba/tests/__init__.py]: It should be quite portable indeed, so `os.path.join` is perhaps mostly a matter of cleanliness or purity :)

**Comment by seibert** [comment]
Can you add a short section to the developer documentation describing how to list the tests?:

https://github.com/numba/numba/blob/master/docs/source/developer/contributing.rst#running-tests

**Comment by stefanseefeld** [comment]
On 12.10.2015 13:30, Stan Seibert wrote:
>
> Can you add a short section to the developer documentation describing
> how to list the tests?:
>
> https://github.com/numba/numba/blob/master/docs/source/developer/contributing.rst#running-tests
>

I have added a patch for that. (Note that this is quite a bit more
substantial than just listing tests. It also makes it easier to run
sub-suites as well as individual tests.)

With this change we may remove the various non-toplevel runtests.py
scripts from sub-suites, as they are now obsolete.

-- 

      ...ich hab' noch einen Koffer in Berlin...

**Comment by seibert** [comment]
Great!  Merging...

**Pull request merged**